### ISSUE TITLE
bootstrap build fails with link error

### ISSUE THREAD

**Issue by dwrensha** [sandstorm-io/ekam#3]
```
$ make
==== compiling bootstrap ekam ====
g++ -Isrc -std=c++11 src/base/Debug.cpp src/base/Promise.cpp src/base/OwnedPtr.cpp src/base/Hash.cpp src/base/sha256.cpp src/os/EpollEventManager.cpp src/os/EventManager.cpp src/os/Subprocess.cpp src/os/File.cpp src/os/ByteStream.cpp src/os/Socket.cpp src/os/DiskFile.cpp src/os/OsHandle.cpp src/os/EventGroup.cpp src/ekam/ExecPluginActionFactory.cpp src/ekam/CppActionFactory.cpp src/ekam/MuxDashboard.cpp src/ekam/initNetworkDashboardStub.cpp src/ekam/Action.cpp src/ekam/Dashboard.cpp src/ekam/ekam.cpp src/ekam/Tag.cpp src/ekam/ConsoleDashboard.cpp src/ekam/Driver.cpp src/ekam/SimpleDashboard.cpp src/ekam/ActionUtil.cpp -pthread -o bin/ekam-bootstrap
==== building ekam with ekam ====
...

link: capnp/compiler/capnpc-c++.o
  tmp/capnp/stringify.o: In function `capnp::(anonymous namespace)::whichFieldType(capnp::StructSchema::Field const&)':
  stringify.c++:(.text+0x4a): undefined reference to `capnp::_::PointerReader::getStruct(capnp::word const*) const'
  tmp/capnp/stringify.o: In function `capnp::_::enumString(unsigned short, capnp::_::RawBrandedSchema const&)':
  stringify.c++:(.text+0xd2): undefined reference to `capnp::_::ListReader::getStructElement(unsigned int) const'
  stringify.c++:(.text+0x10f): undefined reference to `capnp::Text::Reader capnp::_::PointerReader::getBlob<capnp::Text>(void const*, unsigned 
    int) const'
  tmp/capnp/stringify.o: In function `capnp::(anonymous namespace)::print(capnp::DynamicValue::Reader const&, capnp::schema::Type::Which, capnp
    ::(anonymous namespace)::Indent, capnp::(anonymous namespace)::PrintMode)':
  stringify.c++:(.text+0xc64): undefined reference to `capnp::Text::Reader capnp::_::PointerReader::getBlob<capnp::Text>(void const*, unsigned 
    int) const'
  stringify.c++:(.text+0xef2): undefined reference to `capnp::Text::Reader capnp::_::PointerReader::getBlob<capnp::Text>(void const*, unsigned 
    int) const'
  stringify.c++:(.text+0x10e3): undefined reference to `capnp::_::ListReader::getStructElement(unsigned int) const'
  stringify.c++:(.text+0x298d): undefined reference to `capnp::Text::Reader capnp::_::PointerReader::getBlob<capnp::Text>(void const*, unsigned
    int) const'
  tmp/capnp/stringify.o: In function `capnp::prettyPrint(capnp::DynamicStruct::Builder)':
  stringify.c++:(.text+0x35e4): undefined reference to `capnp::_::StructBuilder::asReader() const'
  tmp/capnp/stringify.o: In function `capnp::operator*(kj::_::Stringifier, capnp::DynamicStruct::Builder const&)':
  stringify.c++:(.text+0x38a4): undefined reference to `capnp::_::StructBuilder::asReader() const'
  tmp/capnp/dynamic.o: In function `capnp::DynamicEnum::getEnumerant() const':
  dynamic.c++:(.text+0x3b): undefined reference to `capnp::_::ListReader::getStructElement(unsigned int) const'
  tmp/capnp/dynamic.o: In function `capnp::DynamicList::Builder::asReader() const':
  dynamic.c++:(.text+0x2e4): undefined reference to `capnp::_::ListBuilder::asReader() const'
  tmp/capnp/dynamic.o: In function `capnp::_::PointerHelpers<capnp::DynamicCapability, (capnp::Kind)7>::getDynamic(capnp::_::PointerReader, 
    capnp::InterfaceSchema)':
  dynamic.c++:(.text+0x735): undefined reference to `capnp::_::PointerReader::getCapability() const'
  tmp/capnp/dynamic.o: In function `capnp::_::PointerHelpers<capnp::DynamicCapability, (capnp::Kind)7>::getDynamic(capnp::_::PointerBuilder, 
    capnp::InterfaceSchema)':
  dynamic.c++:(.text+0x770): undefined reference to `capnp::_::PointerBuilder::getCapability()'
  tmp/capnp/dynamic.o: In function `capnp::BuilderFor_<capnp::DynamicStruct, (kind<capnp::DynamicStruct>)()>::Type capnp::Orphan<capnp::
  ...(log truncated; use -l to increase log limit)...
```

It works fine if I revert the capnproto dependency to one commit before https://github.com/sandstorm-io/capnproto/pull/192.

**Comment by kentonv** [comment]
One of the pain points with Ekam is that it has no way of knowing the causality of errors, so the error dump at the end is in an unhelpfully random order. The actual root cause in this case is:

    compile: capnp/layout.c++
      In file included from /ekam-provider/canonical/capnp/layout.c++:24:0:
      /ekam-provider/canonical/capnp/layout.c++: In member function ‘capnp::PointerType capnp::_::PointerBuilder::getPointerType()’:
      /ekam-provider/canonical/capnp/layout.c++:2278:7: error: expected primary-expression before ‘)’ token
             KJ_FAIL_REQUIRE();
             ^
      /ekam-provider/canonical/capnp/layout.c++: In member function ‘capnp::PointerType capnp::_::PointerReader::getPointerType() const’:
      /ekam-provider/canonical/capnp/layout.c++:2385:7: error: expected primary-expression before ‘)’ token
             KJ_FAIL_REQUIRE();
             ^
      /ekam-provider/canonical/capnp/layout.c++: In member function ‘capnp::PointerType capnp::_::PointerBuilder::getPointerType()’:
      /ekam-provider/canonical/capnp/layout.c++:2288:1: warning: control reaches end of non-void function [-Wreturn-type]
       }
       ^
      /ekam-provider/canonical/capnp/layout.c++: In member function ‘capnp::PointerType capnp::_::PointerReader::getPointerType() const’:
      /ekam-provider/canonical/capnp/layout.c++:2395:1: warning: control reaches end of non-void function [-Wreturn-type]
       }
       ^

**Comment by kentonv** [comment]
Apparently the problem *only* occurs when using `g++ -std=c++11`, while Cap'n Proto usually builds using `-std=gnu++0x`.

Fixed in: sandstorm-io/capnproto@959ad29

FWIW, the general trick to finding root causes is: look at compile errors before link errors. (Unfortunately if the capnp tool is failing to build then you'll get a bunch of spurious compile errors because of generated headers not being found.)
